src/vmm/src/device_manager/mmio.rs

```diff
@@ -110,12 +110,12 @@ impl MMIODeviceManager {
             return Err(Error::IrqsExhausted);
         }
 
-        let queue_evts = mmio_device
+        for (i, queue_evt) in mmio_device
```

Comment:
With this change we're keeping the lock on device while registering all events. Is this behavior change on purpose? If yes, can you please specify this in the commit message?

Reply:
No, this is not on purpose, but I didn't expect rust to hold the lock, I should do some experiments and confirm the behavior here.

In case it does hold the lock, I can:
- revert to the explicit syntax used before where the lock is released,
- just leave it like this since during its registration, the device has no other user so there is no contention on the lock whatsoever; nothing really to report in the commit message afaict

Reviewer's choice ^

Reply:
I prefer the previous syntax as there is no need to hold the lock for this operation. Please let me know if your experiments confirm the behavior.

Reply:
I think @andreeaflorescu is right. I've managed to reproduce what she suggests. The conclusion is that the syntax introduced in this PR will keep the lock for the scope of the for loop. How I've reproduced:

```
use std::sync::{Arc, Mutex};
use std::{thread, time};
 
struct HoldsQueueEvts {
     queue_evts: Vec<u32>
}

fn main() {
        let queue_evts = vec![3,2,1];
        let device = Arc::new(Mutex::new(HoldsQueueEvts { queue_evts: queue_evts }));
        let mut handles = Vec::new();
        
        for i in 0..2 {
                let device_clone = device.clone();
                handles.push(thread::spawn(move || {
                        // This thread will block the other from processing.
                        // START
                        //for (idx, evt) in device_clone.lock().unwrap().queue_evts.clone().iter().enumerate() {
                                //println!("{}: {} {}", i, idx, evt);   
                                //if i == 0 {
                                //      let ten_millis = time::Duration::from_millis(2000);
                                //      thread::sleep(ten_millis);
                                //}   
                        //}
                        // END


                        // This is how it wouldn't block the other thread.
                        // START
                        let queue_evts = device_clone.lock().unwrap().queue_evts.clone();
                        for (idx, val) in  queue_evts.iter().enumerate() {
                                println!("{}: {} {}", i, idx, val);
                                if i == 0 {
                                        let ten_millis = time::Duration::from_millis(2000);
                                        thread::sleep(ten_millis);
                                }
                        }
                        // END
                }));
        }

        for handle in handles {
                handle.join().unwrap();
        }
}
```

**OUTPUT 1**:
0: 0 3
0: 1 2
0: 2 1
1: 0 3
1: 1 2
1: 2 1

**OUTPUT 2**:
0: 0 3
1: 0 3
1: 1 2
1: 2 1
0: 1 2
0: 2 1

Reply:
@andreeaflorescu @iulianbarbu 

Holding the lock is actually necessary now because we're taking a reference to the inner object data. In the old code we used to explicitly clone it under the lock then release the lock and provide the independent clone.

The new interface does not force cloning resources anymore. The reference can be safely used while the lock is active. If one would need to release the lock they could do it like before and explicitly clone the data under the reference then release the lock.

In this particular line, I think it's more expensive to clone the data than holding the lock for the duration of the loop, because holding the lock here is free since there is no user user of the device at this time that can suffer from contention.

Reply:
I agree that at this point we do not have any other user of the device. Getting the lock and unlocking is cheap, since it does not imply any syscall, because there is no contention. Also, now we do not clone the queue events anymore, through the interface, but we borrow them, which would imply an explicit cloning step if we want to stick to the old syntax. Furthermore, I think it is not safe to approach this through cloning, since it means that we would have two owners for some specific event fds, which can be closed by one owner, but used still opened by the other, which would not provide shaky consistency for their usage. 

My thought is that we can stick to this syntax at the moment, because theoretically, there will be one single user which registers one specific device, when we build the microVM.